content/JavaScriptCertificationChallenging.q.md

```diff
@@ -0,0 +1,412 @@
+
+```javascript
+// Add your code here...
+
+
+
+
+```
+
+
+/// type=CR, answer=[tests/JavaScriptCertification/PerfectSquareTest.js], difficulty=challenging
+
+Given a positive integer `n`, write a program that finds and returns the least number of perfect squares where their sum is equal to `n`.
+
+```javascript
+// Add your code here...
+
+
+
+
+```
+
+
+/// type=CR, answer=[tests/JavaScriptCertification/CountPrimesTest.js], difficulty=challenging
+
+Given a positive integer `n`, write a program that counts the number of prime numbers less than `n`.
+
+```javascript
+// Add your code here...
+
+
+
+
+```
+
+
+/// type=CR, answer=[tests/JavaScriptCertification/MaxProductSubArrayTest.js], difficulty=challenging
+
+Given an array of integers `arr`, write a program that finds and returns the maximum product of the contiguous subarray within an array that contains at least one number and has the largest product.
+
+```javascript
+// Add your code here...
+
+
+
+
+```
+
+
+/// type=CR, answer=[tests/JavaScriptCertification/MaxAverageSubArrayTest.js], difficulty=challenging
+
+Given an array of integers `arr` of size `n`, write a program that finds and returns the maximum average of the contiguous subarray of the given length `k` within an array that has the largest average. Note: `1 <= k <= n <= 30,000` and the elements of the given array should be in the range of `-10,000` to `10,000`.
+
+```javascript
+// Add your code here...
+
+
+
+
+```
+
+
+/// type=CR, answer=[tests/JavaScriptCertification/TemperatureClassTest.js], difficulty=challenging
+
+Write a complete implementation of a `Temperature` class. The `Temperature` class reads and writes the temperature in either degrees `Celsius` or degrees `Fahrenheit`, but internally it stores only `Celsius` and automatically converts to and from `Celsius` in the Fahrenheit getter and setter. 
+
+```javascript
+// Add your code here...
+
+
+
+
+let temp = new Temperature(37);
+console.log(temp.fahrenheit); // 98.60000000000001
+
+temp.fahrenheit = 108;
+console.log(temp.celsius); //42.22222222222222
+```
+
+
+/// type=CR, answer=[tests/JavaScriptCertification/InsertElementToBodyTest.js], difficulty=challenging
+
+Write a complete implementation of an `insertElement` function that takes two arguments and inserts the specified HTML element with the specified text into the HTML body.
+
+```javascript
+// Add your code here...
+
+
+
+
+```
+
+
+/// type=CR, answer=[tests/JavaScriptCertification/SearchValueInMatrixTest.js], difficulty=challenging
+
+Given a matrix `m x n` with integers in each row are sorted in ascending order from left to right and integers in each column are sorted in ascending order from top to bottom. Write a program with an efficient algorithm that searches for the specified value `x` in an `m x n` matrix. The program returns `true` if `x` is found in the matrix and `false` otherwise.
```

Comment:
I am on the impression for this certification of knowing how good their Javascript coding skills is rather than understanding a complex problem solving. Perhaps we look into how we can break this into easy to understand task but with the same goal of them able to right a Javascript code with an efficient algorithm of finding specified value of `x`.